### ISSUE TITLE
Very strange behaviour w.r.t. num_output

### ISSUE THREAD

**Issue by Deepak-** [BVLC/caffe#3782]
I trained LeNet on the MNIST dataset with `num_output: 10`. This gives me **98%** accuracy.

Then I change`num_output` to `num_output: 5`. This gives **11%** accuracy. The low score makes sense (since there are in fact 10 classes). However a bit surprising that I don't get an error instead, since there is a mismatch.

Anyway, I then split the MNIST dataset in half. So I now I have a _subset_ with classes 0 until 4, and the corresponding validation set. Now I run the same experiment on the subset:

LeNet on the MNIST subset with `num_output: 5` (5, because there's only 5 classes now). This gives me **99%** test accuracy after about 2,000 iterations. 

Here is where it gets weird....

Now I run LeNet again on the subset but with  `num_output: 10`. I again get **99%** test accuracy after 2,000 iterations.

This doesn't make any sense to me since there are only five classes. Could it be a bug?

(also tried it with AlexNet, and gave the same results)

**Comment by shaibagon** [comment]
When trained with `num_output: 5` on dataset with 10 labels you simply tell caffe to read "probabilities" from **wrong** memory addresses. In DEBUG mode there is a check for this, but apparently you are not in debug mode. So, basically you got lucky caffe did not crash on you, but your 11% accuracy is garbage.

When training with `num_output: 10` on dataset of 5 labels basically means there are no gradients and no changes to the last 5 rows of the classification matrix and it is not used. The model learn to ignore these labels as their probability is zero.
